grove/alpha/jordan_gradient/jordan_gradient.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import division
+
+import numpy as np
+import pyquil.quil as pq
+from pyquil.gates import X, H
+from grove.alpha.phaseestimation.phase_estimation import phase_estimation
+
+from grove.alpha.jordan_gradient.gradient_utils import binary_to_real, \
+    measurements_to_bf
+
+
+def initialize_system(ancilla_qubit):
+    """Prepare initial state
+
+    :param list ancilla_qubit: Qubit of ancilla register.
+    :return: Quil program to initialize this system.
+    :rtype: Program
+    """
+
+    # ancilla qubit to plane wave state
+    ic_ancilla = pq.Program([X(ancilla_qubit), H(ancilla_qubit)])
+    p_ic = pq.Program(ic_ancilla)
+
+    return p_ic
+
+def phase_kickback(f_h, precision):
+    """Encode f_h into ancilla eigenvalue and kickback to input registers
+
+    :param float f_h: Oracle output at perturbation h.
+    :param int precision: Bit precision of gradient.
+    :return: Quil program to perform phase kickback.
+    :rtype: Program
+    """
+
+    # encode f_h / 2 into CPHASE gate
+    U = np.array([[np.exp(-1.0j * np.pi * f_h), 0],
```

Comment:
In the paper, Jordan writes that's the phase should actually be `1.j * 2 * np.pi/m * f_h` It looks like the reason this works in the notebook is because you picked the example of 1/2k(x - .5)**2 for x\in[0, 1], so m ends up being 1. (I think) Here `f_h` should be scaled by the range of the gradient of f, so that this works for other spring constants. One way would be to have an input to the function that is the size of the interval of values the gradient can take on. For now, the user can just provide this, which the paper says is reasonable for any numerical algorithm.

I was also puzzled by the fact that the paper says 2pi and you have 1pi here, but that's because he uses a fourier transform, and you're using phase estimation. Am I right in saying that? This code certainly will calculate the gradient, as you've demonstrated, but phase estimation is more expensive than doing the QFT.

In addition, constructing this unitary might be non trivial, and doing a controlled version of this gate in the QFT even more so.

To summarize, I think we should pass a variable that specifies the size of range of values the gradient can take on, and I think we should note in the docstring for this function that we are using quantum phase estimation, rather than the quantum fourier transform, as described in the paper.

Reply:
Scaling for f_h (the m term, now variable `gradient_max`) has been added.  I also updated the example notebook to include this.

So, Jordan's algorithm employs a phase kickback, which is a form of phase estimation, and the end of the phase kickback uses a qft. The end of the `phase_estimation` function currently implemented in Grove also uses a qft.  I've reframed the code to have the 2pi term, using the Grove `phase_estimation` function. It just came down to adjusting the initialization of the ancilla register. In the process, this made the code a bit simpler.